src/kernel/sporadic.c

```diff
@@ -248,94 +252,60 @@ static bool_t refill_head_overlapping(sched_context_t *sc)
 void refill_budget_check(ticks_t usage)
 {
     sched_context_t *sc = NODE_STATE(ksCurSC);
-    /* this function should only be called when the sc is out of budget */
-    ticks_t capacity = refill_capacity(NODE_STATE(ksCurSC), usage);
-    assert(capacity < MIN_BUDGET || refill_full(sc));
-    assert(sc->scPeriod > 0);
+    assert(!isRoundRobin(sc));
     REFILL_SANITY_START(sc);
 
-    if (capacity == 0) {
-        while (refill_head(sc)->rAmount <= usage) {
-            /* exhaust and schedule replenishment */
-            usage -= refill_head(sc)->rAmount;
-            if (refill_single(sc)) {
-                /* update in place */
-                refill_head(sc)->rTime += sc->scPeriod;
-            } else {
-                refill_t old_head = refill_pop_head(sc);
-                old_head.rTime = old_head.rTime + sc->scPeriod;
-                schedule_used(sc, old_head);
-            }
-        }
-
-        /* budget overrun */
-        if (usage > 0) {
-            /* budget reduced when calculating capacity */
-            /* due to overrun delay next replenishment */
-            refill_head(sc)->rTime += usage;
-            /* merge front two replenishments if times overlap */
-            if (!refill_single(sc) &&
-                refill_head(sc)->rTime + refill_head(sc)->rAmount >=
-                refill_index(sc, refill_next(sc, sc->scRefillHead))->rTime) {
-
-                refill_t refill = refill_pop_head(sc);
-                refill_head(sc)->rAmount += refill.rAmount;
-                refill_head(sc)->rTime = refill.rTime;
-            }
-        }
+    /* We assume that the estimates on worst-case kernel overheads are
+     * an upper bound and that overrun cannot occur. */
+#ifdef CONFIG_DEBUG_BUILD
+    if (usage > refill_head(sc)->rAmount) {
+        ticks_t overrun = usage - refill_head(sc)->rAmount;
+        userError(
+            "SC (%p) overran budget by %lluus (%llu ticks)",
+            sc,
+            ticksToUs(overrun),
+            overrun
+        );
     }
-
-    capacity = refill_capacity(sc, usage);
-    if (capacity > 0 && refill_ready(sc)) {
-        refill_split_check(usage);
+#endif
+    usage = MIN(usage, refill_head(sc)->rAmount);
+
+    /* Charge the usage to the head refill */
+    if (usage > 0) {
```

Comment:
IIUC, `usage` could only be zero if we arrive here via `checkBudget` where the current SC is not round-robin but is `refill_full`. Is that right? (If `ksConsumed` was zero, then we should have had `capacity >= MIN_BUDGET` in `checkBudget`.)

Reply:
I think `usage` could only be 0 if `ksConsumed` is still 0 after `updateTimestamp` is called which would indicate getCurrentTime() hasn't returned a value that is strictly larger than a previous value it returned.
So usage > 0 could almost be an assert here I believe.

Reply:
Right, I understand that the `usage` here comes from `ksConsumed` after `updateTimestamp`. I was trying to understand what else must be true to have a zero `usage` here, since I don't think our axiomatisation of time allows us to prove that `getCurrentTime` has advanced. My reasoning was 

In particular, I was looking at `checkBudget`, which calls `chargeBudget` if `capacity < MIN_BUDGET` or if a non-round-robin thread is `refill_full`. My thinking was the `capacity < MIN_BUDGET` can only be true if `ksConsumed` is indeed non-zero, since we maintain the invariant that the head refill is at least `MIN_BUDGET`. And therefore, to have zero `usage` in `refill_budget_check`, we must be a non-round-robin thread which is `refill_full`.

However, there are other calls to `chargeBudget` that I hadn't taken into account, so that reasoning might not hold.

Reply:
I think what you're claiming holds for the other calls to chargeBudget also.  They all either get called with ksConsumed or the head `rAmount` in the case of yield.